### ISSUE TITLE
Revisit string encoding

### ISSUE THREAD

**Issue by PMeira** [dss-extensions/dss_capi#6]
Consider alternative string representations and potential benefits (less conversions, faster copy, etc.)

The current implementation can be kept, e.g., a WideString API would live in parallel with the current AnsiString API.

**Comment by PMeira** [comment]
A first step is to try adding full UTF8 support. We can default to that for input files, optionally adding an option to force to some common encodings. 

`lconvencoding` provides some utilitities but using it probably requires adding Lazarus as a dependency. Since nobody ever complained so far, I guess using UTF8 might be enough.

**Comment by PMeira** [comment]
I'll probably work on this along side with #93.

Besides code sections that address strings by index and/or use the number of bytes == length assumption, UTF8 should work without further changes. It's important to add a note in some properties/commands/functions where only ASCII are accepted, e.g. the parser functions:
https://github.com/dss-extensions/dss_capi/blob/0ef72db7eb046ad5d7a3731ff5ea33b8089b728c/include/dss_capi.h#L3285-L3304

**Potential optimization in the language bindings**

(For UTF8, all these will still be valid.)

Since Pascal strings are prepended by their lengths, there is a simple optimization available there -- for simple/single Pascal strings there is a global buffer in DSS C-API, so we don't need to de-allocate them as an API user. In this case, the original data (length, refcount) is still there. 

This is an implementation detail of the FPC string though. If we change/mix languages later, it might not be the best to advertise, but we could use that for the projects we control.

I just did a simple benchmark using Cython (CFFI in CPython would be slower) and avoiding strlen makes the performance better and more stable. 

For arrays of strings, a faster/better API than returning the arrays of chars would be preferable. We can also change just change the current `DSS_CopyStringAsPChar`:

https://github.com/dss-extensions/dss_capi/blob/0ef72db7eb046ad5d7a3731ff5ea33b8089b728c/src/CAPI/CAPI_Utils.pas#L218-L222

- This uses `StrCopy` (null-determined strings), which I didn't pay attention before. I'm changing it to use `StrPCopy` now (things get a few percent points faster), which also uses the intrinsic length. 
- We can over-allocate to include the length (1 native integer, so 8 bytes for x64). I can test and benchmark this in a later date. This would change `DSS_CopyStringAsPChar` again (just use a plain `Move`/memcpy instead of ` StrPCopy` ) and `DSS_Dispose_PPAnsiChar` to address the offset pointers.

I tested with Cython but this should apply to Julia, C++, and Matlab when using MEX down the line.
